### PR TITLE
fix(redaction): mask a KEY=value assignment inside a quoted string leaf (#178)

### LINKED ISSUE #178: Display redaction misses a KEY=value assignment inside a quoted JSON string leaf
## Summary

Split out of #169 as promised in https://github.com/JeongJaeSoon/agent-guard/issues/169#issuecomment-5470137487. Found while comparing v3.0.1 and `main` for that issue; it is a distinct root cause from both #169 (over-masking, fixed by `ff50c21`) and #156 (status-label allowlist).

The env-assignment heuristic does not mask a `KEY=value` assignment when the assignment sits **inside a double-quoted string** — which is the ordinary shape of an MCP tool response that returns a serialized JSON log. Reproduced on current `main` (`818f540`) via `hook-post-tool`.

```
IN : {"level":"info","log":"starting with API_KEY=hunter2-long-value","ok":true}
OUT: (no rewrite — the value reaches the model in full)
```

`hunter2-long-value` is the value `tests/run.sh` already uses to stand for "a secret gitleaks misses", so this is the heuristic path specifically. Anything gitleaks recognises is still masked in the same shape, which is what pins it:

```
IN : {"log":"AGENT_GUARD_TEST_SECRET here"}   ->  {"log":"[REDACTED] here"}   # gitleaks path, fine
```

## Isolation

The trigger is the string's own **closing quote**, not the nesting or the JSON:

```
DB_PASSWORD=hunter2-long-value                    ->  DB_PASSWORD=[REDACTED]
 "DB_PASSWORD=hunter2-long-value                  ->   "DB_PASSWORD=[REDACTED]      # no closing quote
 "DB_PASSWORD"=hunter2-long-value                 ->   "DB_PASSWORD"=[REDACTED]     # quoted KEY is fine
prefix DB_PASSWORD=hunter2-long-value suffix      ->  prefix DB_PASSWORD=[REDACTED] suffix
prefix "DB_PASSWORD=hunter2-long-value" suffix    ->  (no rewrite)                  # <- the gap
{"log":"DB_PASSWORD=hunter2-long-value"}          ->  (no rewrite)
{"log":"DB_PASSWORD=hunter2-long-value,x"}        ->  (no rewrite)
```

For contrast, the **colon** form in the same JSON masks correctly, because it reaches `emit_value`'s quoted branch instead:

```
{"DB_PASSWORD":"hunter2-long-value"}              ->  {"DB_PASSWORD":"[REDACTED]"}
```

## Root cause

`equals_assignment` matches fine. The value token then keeps the string's closing quote — `hunter2-long-value"` — because the unquoted branch only cuts at `[ \t,;]` and strips trailing `}`, `]`, `)`. That token then hits the embedded-quote test in `looks_like_reference()` (`plugins/agent-guard/bin/agent-guard:3256-3259`):

```awk
if (index(value, "(") > 1 || index(value, "[") > 1 \
    || index(value, "{") > 1 || index(value, "\"") \
    || index(value, "'") || index(value, "`"))
  return 1
```

`index(value, "\"")` is unpositioned, so a quote **anywhere** — including as the final character — marks the token a reference and it is skipped.

The test itself is doing real work: it is what keeps `password = os.environ["DB_PASSWORD"]` and friends from being mangled (the #169 source-code assertions). The code comment already records the tradeoff — *"an unquoted credential that itself contains an opener or an embedded quote is no longer masked by this heuristic"* — but I don't think the intent was to include the whole class of *assignments inside a JSON string leaf*, which is the most common MCP tool-response shape rather than an edge case.

## Suggested direction

Position the quote test the way the opener tests are already positioned. An opener only counts past the first character; by the same logic a quote that is the **last** character is the enclosing string's terminator, not a nested string:

- treat a quote as embedded only when it occurs before the final character;
- strip that single trailing quote from the value before eligibility and replacement, the way trailing closers are already stripped, so the replacement extent stays exactly the value.

`os.environ["DB_PASSWORD"]` keeps its interior quotes and stays a reference, so the source-code assertions should hold. Worth checking the odd/even case too (`"a"b"`), and whether the same positioning should apply to `'` and `` ` ``.

## Acceptance

- must-mask: `{"log":"API_KEY=<value>"}`, `prefix "DB_PASSWORD=<value>" suffix`, and the nested MCP `content[].text` form on both the Claude and Codex contracts;
- must-not-mask: every existing source-code reference assertion from #169 (`process.env.API_KEY`, `os.environ["DB_PASSWORD"]`, `config.apiKey`, `ENV.fetch("SESSION_KEY")`, `${API_KEY:?required}`);
- the replacement extent stays exactly the value — the enclosing quote and the rest of the leaf are preserved, per the #169 regression test added in https://github.com/JeongJaeSoon/agent-guard/pull/177;
- block/detect path unchanged; this is PostToolUse display redaction only.

## Non-goals

- No change to the gitleaks path, which already covers this shape.
- Do not remove the reference heuristic — that re-breaks the source-code preservation #169 is about.

### PR DESCRIPTION
**Pull request by JeongJaeSoon**
Closes #178

Split out of #177 as promised there. Reproduced on `main` before and after that merge — #177 does not touch this path.

## The gap

The ordinary shape of a tool that returns a serialized log reached the model in full, on the heuristic path:

```
IN : {"level":"info","log":"starting with API_KEY=hunter2-long-value","ok":true}
OUT: (no rewrite)
```

`hunter2-long-value` is the value the suite already uses to stand for "a secret gitleaks misses", so this is the heuristic specifically — anything gitleaks recognises was masked in the same shape throughout, which is what pins it.

## Two causes, both required

I expected one. Fixing only the quote left the JSON form still unmasked, which is what turned up the second.

**1. The enclosing string's closing quote rode along on the value token.** The unquoted branch cuts at `[ \t,;]` and strips trailing `}`, `]`, `)`, so the token was `hunter2-long-value"`. `looks_like_reference()` then tests for an embedded quote with an *unpositioned* `index()`:

```awk
|| index(value, "\"") || index(value, "'") || index(value, "`"))
  return 1
```

A quote anywhere — including as the final character — marks the token a nested reference and skips it.

Fixed the way the issue suggests, and the way the opener tests are already positioned: strip exactly one trailing quote, only at the very end, so an interior quote still marks a reference.

With one guard the issue does not mention. The strip must **not** fire when the assignment itself begins with a quote. That marks the line-start re-scan of a value an outer quoted assignment already claimed: `scan_line` advances past the matched key only, so the inner text is examined a second time and matches via the `^` alternative. The outer match has already emitted the whole quoted value; stripping again produced a second, narrower mapping that beat it and stranded the inner key prefix on display:

```
IN  : PASSWORD=abc) API_TOKEN="PASSWORD=abc)" status=ok
BAD : PASSWORD=abc) API_TOKEN="PASSWORD=[REDACTED])" status=ok   <- key prefix visible
NOW : PASSWORD=abc) API_TOKEN="[REDACTED]" status=ok
```

That case is pinned as a test. The guard has a cost, tracked separately in #180: a leaf line that *begins* with a quoted assignment still leaks. That shape leaks on `main` too, so it is residual rather than a regression, and getting both properties needs a different change from this one.

**2. `:` was not a delimiter before a quoted key.** `equals_assignment` required `(^|[ \t,{;([])` before the optional opening quote. In a `{"log":"KEY=<value>"}` shape the only thing between the structure and the key is `:`, so the entire class went unmatched regardless of the quote fix.

It must sit **away from `[`** in the bracket expression — `[:` opens a POSIX character class and invalidates the whole regex, which silently disables *all* masking rather than failing loudly. Confirmed the hard way in review: `[ \t,{;([:]` is fatal in both awks (`mawk: bad class`, `gawk: Unmatched [, [^, [:`), while `[ \t,:{;([]` is clean under `gawk --lint` and `--posix`.

## Two defects found in adversarial review, both fixed

Written up in full [here](https://github.com/JeongJaeSoon/agent-guard/pull/179#issuecomment-5471887931). Both were in the strip loop, and both would have shipped without the review.

**`f2ae9f8` — the strip walked into the redaction sentinel.** With a quote after it, `password=[REDACTED]"}` lost `}`, then the quote, then the `]` belonging to the marker. `already_redacted()` ran only *before* the loop, so `[REDACTED` was re-emitted as a secret and a `]` appended — compounding on every pass, damaging every `[REDACTED` in the response, taking the prompt guard from exit 0 to **exit 2** on text Agent Guard had itself redacted, and annihilating a >64 KiB non-UTF-8 `exec` stream that contained nothing else. Now re-tested after every strip.

**`985f4f1` — one carriage return defeated the whole fix.** `{"log":"API_KEY=<secret>"}\r\n` did not mask at all, while the LF form did: the loop stopped on the CR before reaching the quote. That is Windows tools, PowerShell, Docker and CI logs. `\r` is edge whitespace here, as it already is in `edge_trim()`; the extent still stops at the value, so the carriage return survives in the output.

## Verification

Acceptance criteria from the issue, run through the real hook binary:

| | |
|---|---|
| `{"log":"API_KEY=…"}` | masked |
| `{"level":"info","log":"starting with API_KEY=…","ok":true}` | masked, siblings and JSON shape preserved |
| `prefix "DB_PASSWORD=…" suffix` | masked |
| `{"a":{"b":"API_KEY=…"}}` | masked |
| `{"log":"DB_PASSWORD=…,x"}` | masked, `,x` preserved |
| the same with CRLF line endings | masked, the CR preserved |
| nested MCP `content[].text`, both host contracts | masked |
| clean nested response | no rewrite |
| a leaf already carrying `[REDACTED]` | no rewrite, idempotent |
| the same shape at the prompt guard | rc 2 for a secret, rc 0 for a sentinel |

Must-not-mask, the #169 source-code preservation assertions, all still pass unchanged:

```
process.env.API_KEY          preserved
os.environ["DB_PASSWORD"]    preserved
config.apiKey                preserved
ENV.fetch("SESSION_KEY")     preserved
${API_KEY:?required}         preserved
```

- Differential against a real `94af279` worktree: **5,252 cases (1,252 structured + 4,000 fuzz), 0 regressions**, checked by extracting every span the baseline replaced and failing if it survives verbatim — not by marker presence. `hook-user-prompt` exit codes compared across the corpus: no case where base blocks and head allows.
- `tests/run.sh`: **1107 passed, 0 failed**, identical under mawk 1.3.4 and gawk 5.2.1.
- `shellcheck --severity=error` clean; plugin scanner **0 high, 96/100** (plugin-scanner 2.2.121, matching the CI action).
- Replacement extent is exactly the value: the enclosing quote, the carriage return and the rest of the leaf are preserved, per the #169 regression test merged in #177.
- Block/detect path unchanged. This is display redaction plus the prompt guard, which shares the same heuristic.

## Considered and dismissed, with measurements

Two review findings turned out to be pre-existing behaviour that this PR extends consistently rather than new bugs — the reasoning and numbers are in the review comment. In short: global masking of an ordinary word (`api_key=production` scrubbing every `production`) is what `main` already does for the unquoted form, and total replacement of a large non-UTF-8 `exec` output is the raw probe path being all-or-nothing by design. Making the JSON shape quieter than the bare one would be the inconsistency.

## Non-goals honoured

The gitleaks path is untouched, and the reference heuristic is not removed — only positioned.

### CODE REVIEW

**Comment by JeongJaeSoon** [comment]
## Adversarial review round 1 — two real defects, both fixed; two dismissed with evidence

An independent adversarial differential pass against a real `94af279` worktree found five things. I reproduced all five before acting. Two were genuine and are fixed in `f2ae9f8` and `985f4f1`; two are pre-existing behaviour that this PR merely extends consistently, with measurements below; one is a residual leak now filed as #180.

The headline: the reviewer was right that the change as it stood **should not have merged**, and it was not the `:` widening or the regex — it was the strip walking one character too far.

### Fixed 1 — the strip walked into the redaction sentinel

The worst finding, and worse than the bug this PR set out to fix.

```
IN  : {"log":"password=[REDACTED]"}
BASE: (no rewrite)
WAS : {"log":"password=[REDACTED]]"}
NOW : (no rewrite)
```

With a quote after it — exactly the leaf shape this PR adds — the loop took `}`, then the quote, then the `]` that belongs to the marker. `already_redacted()` runs *before* the loop, where the tail still held a quote, so it passed; afterwards `[REDACTED` matches nothing, so the fragment was re-emitted as a secret and a `]` appended. It compounded on every pass (`[REDACTED]]` → `[REDACTED]]]` → …), the replacement is global so every `[REDACTED` in the response was damaged, and on the prompt guard the same input went from exit 0 to **exit 2** — Agent Guard blocking text it had itself redacted. On `agent-guard exec` over a >64 KiB invalid-UTF-8 stream the probe path replaces the whole output, so a 70 KB log containing nothing but a prior sentinel was annihilated.

Fix: re-test `already_redacted()` after every strip, not only before. The existing idempotence assertions cover only the bare `PASSWORD=[REDACTED]` form with no trailing quote, which is why they stayed green; the quoted-leaf, nested, multi-marker and prompt-guard shapes are covered now.

### Fixed 2 — one carriage return defeated the whole fix

```
{"log":"API_KEY=<secret>"}\n     ->  {"log":"API_KEY=[REDACTED]"}
{"log":"API_KEY=<secret>"}\r\n   ->  no rewrite, secret in full
```

`scan_leaf` splits on `\n` and keeps the preceding `\r`, so the token was `…"}\r` and the loop stopped on the CR before reaching the quote. That is Windows tools, PowerShell, Docker and CI logs — not an edge case. `edge_trim()` already treats `\r` as edge whitespace everywhere else; this loop was the one place that did not. The replacement extent still stops at the value, so the carriage return survives in the rewritten text — asserted, since that extent guarantee is what #169 pins.

### Dismissed — global over-masking of an ordinary word is pre-existing

The reviewer flagged `{"log":"using api_key=production","env":"production"}` scrubbing both `production`s. That is not new; it is what `main` already does for the unquoted form, and this PR makes the JSON-wrapped shape behave the same:

```
IN   : using api_key=production and region production
BASE : using api_key=[REDACTED] and region [REDACTED]
HEAD : using api_key=[REDACTED] and region [REDACTED]
```

Same for `password=localhost`, `token=v1.2.3`. The code comment at the `emit_secret` site already states the tradeoff and gives `response: token: v1.2.3` as the precedent. Making one shape quieter than the others would be the inconsistency, so this stays.

### Dismissed — total replacement of a large non-UTF-8 output is pre-existing

`agent-guard exec` replaces the entire output when the raw probe fires past 64 KiB. The reviewer showed a 70 KB binary annihilated by a benign `api_key=production`. Measured against base with the *bare* form:

```
70 KB + \xff\xfe + 'api_key=production'            BASE: 11 bytes   HEAD: 11 bytes
70 KB + \xff\xfe + '{"log":"api_key=production"}'  BASE: 70040      HEAD: 11 bytes
```

Base already annihilates it for the unquoted form — the probe path is all-or-nothing by design, and this PR only makes the JSON shape reach it. The one row of that finding that *was* a new bug is the sentinel case, which is Fixed 1 above; that row now returns 70041 bytes on both.

### Filed — the residual leak

A leaf line that *begins* with a quoted assignment (`"API_KEY=<secret>"`) still leaks, on `main` and on this branch alike. The guard that prevents it is load-bearing: removing it strands a key prefix on display in a case the suite already pins. Getting both requires testing the "this is the outer capture" property directly rather than inferring it from the first character, which is a different change from this one. Filed as #180 with the reproduction, the measurement that proves the guard is the sole cause, and acceptance criteria that force both properties at once. Not widening this PR for it.

### Verification on `985f4f1`

- Differential vs a real `94af279` worktree: **5,252 cases (1,252 structured + 4,000 fuzz), 0 regressions** — checked by extracting every span the baseline replaced and failing if it survives verbatim, not by marker presence. The single flagged hit is the key name `PASSWORD=`, where this branch has the *better* extent (base over-replaced the key, head masks only the value).
- `hook-user-prompt` exit codes across the corpus: no case where base blocks and head allows.
- `tests/run.sh`: **1107 passed, 0 failed**, identical under mawk 1.3.4 and gawk 5.2.1.
- `shellcheck --severity=error` clean; plugin scanner 0 high, 96/100.
- Both awks confirmed on the regex claim from the commit message: the naive placement `[ \t,{;([:]` is fatal in both (`mawk: bad class`, `gawk: Unmatched [, [^, [:`), and `[ \t,:{;([]` is clean under `gawk --lint` and `--posix`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: plugins/agent-guard/bin/agent-guard [hunk 1]
```diff
@@ -3026,10 +3026,15 @@ secretish_env_values() {
...
+      # begin immediately after one (`wrap(secret_key=<secret>)`). `:` is here
+      # for the serialized-log leaf a tool returns as JSON: in a
+      # `{"log":"KEY=<value>"}` shape the only thing between the structure and
+      # the key is `:`, so without it the whole class went unmatched (#178).
+      # It must sit away from `[` in the class — `[:` opens a POSIX character
+      # class and makes the whole regex invalid. The value gate
       # below stops treating a whole call expression as a single literal, so a
       # nested assignment has to be reachable on its own to stay masked.
-      equals_assignment = "(^|[ \t,{;([])[\"'\''`]?" key "[\"'\''`]?[ \t]*="
+      equals_assignment = "(^|[ \t,:{;([])[\"'\''`]?" key "[\"'\''`]?[ \t]*="
```
[the remark this answers is not in the document]
Comment (JeongJaeSoon) [reply plugins/agent-guard/bin/agent-guard:3037]: Reproduced, and I am not taking it — the behaviour delta is real but the premise is not. Apologies for the timing: I marked the PR ready and merged it before this review landed, which was my mistake. Measurements below; if they do not convince you, this reopens as a follow-up issue.

**The delta is real.** Against the parent (`94af279`):

```
label:API_KEY=config.apiKey;      BASE: no rewrite, prompt rc 0   HEAD: [REDACTED], prompt rc 2
```

**But reference detection never applied to this form.** The dotted-chain rule in `looks_like_reference()` is deliberately gated on `assignment ~ /[ \t]=$/` — whitespace on both sides of `=`, a "code-shaped assignment". Your example is unspaced, so it falls outside that gate, and it always has. The parent masks the identical unspaced assignment in every other position:

```
API_KEY=config.apiKey;            BASE: API_KEY=[REDACTED];   prompt rc 2   <- parent, no colon involved
API_KEY=process.env.API_KEY;      BASE: API_KEY=[REDACTED];   prompt rc 2   <- parent
```

So the parent did not preserve `label:API_KEY=config.apiKey;` because it recognised a reference. It preserved it because `:` was not a delimiter, so the assignment never matched at all. Adding `:` makes the labelled position behave like every other position, rather than removing a protection.

The protection that does exist is intact, labelled or not:

```
api_key = config.apiKey;          BASE: no rewrite   HEAD: no rewrite
api_key = process.env.API_KEY;    BASE: no rewrite   HEAD: no rewrite
label:api_key = config.apiKey;    BASE: no rewrite   HEAD: no rewrite
label:api_key = process.env.API_KEY;  BASE: no rewrite   HEAD: no rewrite
```

That spaced form is what an assignment in real JS/TS source looks like, and it is what the #169 preservation assertions pin. All five of them still pass unchanged.

**What your narrowing would cost.** Requiring a quote after `:` would restore the unspaced labelled statement and give back this, which the parent leaks and the merged change masks:

```
{log:API_KEY=hunter2-long-value}          BASE: no rewrite   HEAD: {log:API_KEY=[REDACTED]}
cfg{db:password=hunter2-long-value}       BASE: no rewrite   HEAD: cfg{db:password=[REDACTED]}
trace:token=hunter2-long-value            BASE: no rewrite   HEAD: trace:token=[REDACTED]
```

Unquoted `:`-delimited assignments are ordinary in JS object literals and structured log formats. Trading three real leaks for one display false positive on a labelled statement written without spaces runs against the tradeoff this path is built on — the comment at the `emit_secret` site states it, and `main` already scrubs `v1.2.3` from a whole response for `response: token: v1.2.3`.

**Where I think you are right, and it is a bigger fish.** The unspaced/spaced split is doing more work than it should: `API_KEY=config.apiKey` masking while `api_key = config.apiKey` does not is a thin distinction, and your example is a fair illustration of that. But it predates this PR, it is not specific to `:`, and narrowing `:` would only hide one instance of it while costing coverage. If you want that revisited it should be its own issue against `looks_like_reference()`, and I would rather see it argued on the spacing gate than on the delimiter.

Full review record for this PR, including two defects an adversarial pass caught and I fixed before merge, is in [this comment](https://github.com/JeongJaeSoon/agent-guard/pull/179#issuecomment-5471887931).

---
_Generated by [Claude Code](https://claude.ai/code)_
